presto-main/src/main/java/com/facebook/presto/execution/QueryExecution.java

```diff
@@ -59,6 +59,8 @@ Duration waitForStateChange(QueryState currentState, Duration maxWait)
 
     void addStateChangeListener(StateChangeListener<QueryState> stateChangeListener);
 
+    void addQueryInfoStateChangeListener(StateChangeListener<QueryInfo> stateChangeListener);
```

Comment:
I'd name this `addFinalQueryInfoListener` and document that it will only be called once with a final query info.